// thinkjs/lib/adapter/db/mongo.js
'use strict';

exports.__esModule = true;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _parse_mongo = require('./_parse_mongo.js');

var _parse_mongo2 = _interopRequireDefault(_parse_mongo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MongoSocket = think.adapter('socket', 'mongo');

/**
 * mongo db class
 */

var _class = function (_Parse) {
  (0, _inherits3.default)(_class, _Parse);

  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    return (0, _possibleConstructorReturn3.default)(this, _Parse.apply(this, arguments));
  }

  /**
   * init
   * @param  {Object} config []
   * @return {}        []
   */
  _class.prototype.init = function init(config) {
    _Parse.prototype.init.call(this);
    this.config = config;
    this.lastInsertId = 0;
    this._socket = null; //Mongo socket instance
  };
  /**
   * connect mongo socket
   * @return {Promise} []
   */


  _class.prototype.socket = function socket() {
    if (this._socket) {
      return this._socket;
    }
    this._socket = MongoSocket.getInstance(this.config, thinkCache.DB);
    return this._socket;
  };
  /**
   * get connection
   * @return {Promise} []
   */


  _class.prototype.collection = function collection(table) {
    var instance = this.socket();
    return instance.getConnection().then(function (db) {
      return db.collection(table);
    });
  };
  /**
   * get last insert id
   * @return {String} []
   */


  _class.prototype.getLastInsertId = function getLastInsertId() {
    return this.lastInsertId;
  };
  /**
   * add data
   * @param {Objec} data    []
   * @param {Object} options []
   */


  _class.prototype.add = function () {
    var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(data, options) {
      var collection, result;
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context.sent;
              _context.next = 5;
              return collection.insert(data);

            case 5:
              result = _context.sent;

              this.lastInsertId = data._id.toString();
              return _context.abrupt('return', result);

            case 8:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    }));

    function add(_x, _x2) {
      return _ref.apply(this, arguments);
    }

    return add;
  }();
  /**
   * add multi data
   * @param {Array} dataList []
   * @param {Object} options [ {ordered: true}. If false, perform an unordered insert, and if an error occurs with one of documents, continue processing the remaining documents in the array.}]
   * @param {Object} options  []
   */


  _class.prototype.addMany = function () {
    var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(dataList, options) {
      var collection, result, insertedIds;
      return _regenerator2.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context2.sent;
              _context2.next = 5;
              return collection.insert(dataList, options);

            case 5:
              result = _context2.sent;
              insertedIds = dataList.map(function (item) {
                return item._id.toString();
              });

              this.lastInsertId = insertedIds;
              return _context2.abrupt('return', result);

            case 9:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, this);
    }));

    function addMany(_x3, _x4) {
      return _ref2.apply(this, arguments);
    }

    return addMany;
  }();
  /**
   * set collection limit
   * @param  {Object} collection []
   * @param  {String} limit      []
   * @return {Object}            []
   */


  _class.prototype.limit = function limit(collection, _limit) {
    _limit = this.parseLimit(_limit);
    if (_limit[0]) {
      collection.skip(_limit[0]);
    }
    if (_limit[1]) {
      collection.limit(_limit[1]);
    }
    return collection;
  };
  /**
   * parse group
   * @param  {String} group []
   * @return {Object}       []
   */


  _class.prototype.group = function group(_group) {
    _group = this.parseGroup(_group);
    var length = _group.length;
    if (length === 0) {
      return { _id: null };
    } else if (length === 1) {
      return { _id: '$' + _group[0] };
    } else {
      var _ret = function () {
        var result = {};
        _group.forEach(function (item) {
          result[item] = '$' + item;
        });
        return {
          v: result
        };
      }();

      if ((typeof _ret === 'undefined' ? 'undefined' : (0, _typeof3.default)(_ret)) === "object") return _ret.v;
    }
  };
  /**
   * select data
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.select = function () {
    var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(options) {
      var collection, where, distinct;
      return _regenerator2.default.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context3.sent;
              where = this.parseWhere(options.where);

              //get distinct field data

              distinct = this.parseDistinct(options.distinct);

              if (!distinct) {
                _context3.next = 7;
                break;
              }

              return _context3.abrupt('return', collection.distinct(distinct, where));

            case 7:

              collection = collection.find(where, this.parseField(options.field));
              collection = this.limit(collection, options.limit);
              collection = collection.sort(this.parseOrder(options.order));
              return _context3.abrupt('return', collection.toArray());

            case 11:
            case 'end':
              return _context3.stop();
          }
        }
      }, _callee3, this);
    }));

    function select(_x5) {
      return _ref3.apply(this, arguments);
    }

    return select;
  }();
  /**
   * update data
   * @param  {Object} data    []
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.update = function () {
    var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(data, options) {
      var collection, where, limit, flag, key;
      return _regenerator2.default.wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context4.sent;
              where = this.parseWhere(options.where);
              limit = this.parseLimit(options.limit);
              // updates multiple documents that meet the query criteria. 
              // default only updates one document

              if (limit[1] !== 1) {
                options.multi = true;
              }

              // If set to true, creates a new document when no document matches the query criteria. 
              // The default value is false, which does not insert a new document when no match is found.
              if (!options.upsert) {
                options.upsert = false;
              }

              //add $set for data
              flag = true;
              _context4.t0 = _regenerator2.default.keys(data);

            case 9:
              if ((_context4.t1 = _context4.t0()).done) {
                _context4.next = 16;
                break;
              }

              key = _context4.t1.value;

              if (!(key[0] !== '$')) {
                _context4.next = 14;
                break;
              }

              flag = false;
              return _context4.abrupt('break', 16);

            case 14:
              _context4.next = 9;
              break;

            case 16:
              if (!flag) {
                data = { $set: data };
              }

              // update operator
              // http://docs.mongodb.org/manual/reference/operator/update/#id1
              return _context4.abrupt('return', collection.update(where, data, options));

            case 18:
            case 'end':
              return _context4.stop();
          }
        }
      }, _callee4, this);
    }));

    function update(_x6, _x7) {
      return _ref4.apply(this, arguments);
    }

    return update;
  }();
  /**
   * delete data
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.delete = function () {
    var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(options) {
      var collection, where, limit, removeOpt;
      return _regenerator2.default.wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context5.sent;
              where = this.parseWhere(options.where);
              limit = this.parseLimit(options.limit);

              //delete one row

              removeOpt = {};

              if (limit[1] === 1) {
                removeOpt.justOne = true;
              }

              return _context5.abrupt('return', collection.remove(where, removeOpt));

            case 8:
            case 'end':
              return _context5.stop();
          }
        }
      }, _callee5, this);
    }));

    function _delete(_x8) {
      return _ref5.apply(this, arguments);
    }

    return _delete;
  }();
  /**
   * get count
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.count = function () {
    var _ref6 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6(options) {
      var collection, where, group, order, aggregate, fn;
      return _regenerator2.default.wrap(function _callee6$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context6.sent;
              where = this.parseWhere(options.where);
              group = this.group(options.group);

              group.total = { $sum: 1 };

              order = this.parseOrder(options.order);
              aggregate = [];

              if (!think.isEmpty(where)) {
                aggregate.push({ $match: where });
              }
              aggregate.push({ $group: group });
              if (!think.isEmpty(order)) {
                aggregate.push({ $sort: order });
              }
              //make aggregate method to be a promise
              fn = think.promisify(collection.aggregate, collection);
              return _context6.abrupt('return', fn(aggregate).then(function (data) {
                return data[0] && data[0].total || 0;
              }));

            case 13:
            case 'end':
              return _context6.stop();
          }
        }
      }, _callee6, this);
    }));

    function count(_x9) {
      return _ref6.apply(this, arguments);
    }

    return count;
  }();
  /**
   * get sum
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.sum = function () {
    var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(options) {
      var collection, where, group, order, aggregate, fn;
      return _regenerator2.default.wrap(function _callee7$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return this.collection(options.table);

            case 2:
              collection = _context7.sent;
              where = this.parseWhere(options.where);
              group = this.group(options.group);

              group.total = { $sum: '$' + options.field };

              order = this.parseOrder(options.order);
              aggregate = [];

              if (!think.isEmpty(where)) {
                aggregate.push({ $match: where });
              }
              aggregate.push({ $group: group });
              if (!think.isEmpty(order)) {
                aggregate.push({ $sort: order });
              }
              //make aggregate method to be a promise
              fn = think.promisify(collection.aggregate, collection);
              return _context7.abrupt('return', fn(aggregate).then(function (data) {
                return data[0] && data[0].total || 0;
              }));

            case 13:
            case 'end':
              return _context7.stop();
          }
        }
      }, _callee7, this);
    }));

    function sum(_x10) {
      return _ref7.apply(this, arguments);
    }

    return sum;
  }();
  /**
   * create collection indexes
   * @param  {String} table   []
   * @param  {Object} indexes []
   * @return {Promise}         []
   */


  _class.prototype.ensureIndex = function ensureIndex(table, indexes) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

    if (options === true) {
      options = { unique: true };
    }
    if (think.isString(indexes)) {
      indexes = indexes.split(/\s*,\s*/);
    }
    if (think.isArray(indexes)) {
      (function () {
        var result = {};
        indexes.forEach(function (item) {
          result[item] = 1;
        });
        indexes = result;
      })();
    }
    return this.collection(table).then(function (collection) {
      return collection.ensureIndex(indexes, options);
    });
  };
  /**
   * aggregate
   * @param  {String} table   []
   * @param  {Object} options []
   * @return {Promise}         []
   */


  _class.prototype.aggregate = function aggregate(table, options) {
    return this.collection(table).then(function (collection) {
      var fn = think.promisify(collection.aggregate, collection);
      return fn(options);
    });
  };
  /**
   * close socket
   * @return {} []
   */


  _class.prototype.close = function close() {
    if (this._socket) {
      this._socket.close();
      this._socket = null;
    }
  };

  return _class;
}(_parse_mongo2.default);

exports.default = _class;
