// thinkjs/lib/adapter/db/postgresql.js
'use strict';

exports.__esModule = true;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _base = require('./base.js');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PostgreSocket = think.adapter('socket', 'postgresql');
/**
 * postgre db
 */

var _class = function (_Base) {
  (0, _inherits3.default)(_class, _Base);

  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    return (0, _possibleConstructorReturn3.default)(this, _Base.apply(this, arguments));
  }

  /**
   * init
   * @return {} []
   */
  _class.prototype.init = function init() {
    var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    this.config = config;
    //operate
    this.comparison = {
      'EQ': '=',
      'NEQ': '!=',
      '<>': '!=',
      'GT': '>',
      'EGT': '>=',
      'LT': '<',
      'ELT': '<=',
      'NOTLIKE': 'NOT LIKE',
      'LIKE': 'LIKE',
      'NOTILIKE': 'NOT ILIKE',
      'ILIKE': 'ILIKE',
      'IN': 'IN',
      'NOTIN': 'NOT IN',
      'BETWEEN': 'BETWEEN',
      'NOTBETWEEN': 'NOT BETWEEN'
    };
    this.selectSql = '%EXPLAIN%SELECT%DISTINCT% %FIELD% FROM %TABLE%%JOIN%%WHERE%%GROUP%%HAVING%%ORDER%%LIMIT%%UNION%%COMMENT%';
  };
  /**
   * get postgre socket instance
   * @param  {Object} config []
   * @return {}        []
   */


  _class.prototype.socket = function socket(sql) {
    if (this._socket) {
      return this._socket;
    }
    var config = think.extend({
      sql: sql
    }, this.config);
    this._socket = PostgreSocket.getInstance(config, thinkCache.DB, ['sql']);
    return this._socket;
  };
  /**
   * get table info
   * @param  {String} table [table name]
   * @return {Promise}       []
   */


  _class.prototype.getSchema = function () {
    var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(table) {
      var columnSql, columnsPromise, indexSql, indexPromise, _ref2, columns, indexs, schema, extra, reg;

      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              columnSql = 'SELECT column_name,is_nullable,data_type FROM INFORMATION_SCHEMA.COLUMNS WHERE table_name=\'' + table + '\'';
              columnsPromise = this.query(columnSql);
              indexSql = 'SELECT indexname,indexdef FROM pg_indexes WHERE tablename=\'' + table + '\'';
              indexPromise = this.query(indexSql);
              _context.next = 6;
              return _promise2.default.all([columnsPromise, indexPromise]);

            case 6:
              _ref2 = _context.sent;
              columns = _ref2[0];
              indexs = _ref2[1];
              schema = {};

              columns.forEach(function (item) {
                schema[item.column_name] = {
                  name: item.column_name,
                  type: item.data_type,
                  required: item.is_nullable === 'NO',
                  default: '',
                  auto_increment: false
                };
              });
              extra = {};
              reg = /\((\w+)(?:, (\w+))*\)/;

              indexs.forEach(function (item) {
                var _item$indexdef$match = item.indexdef.match(reg),
                    name = _item$indexdef$match[1],
                    others = _item$indexdef$match.slice(2);

                extra[name] = {};
                if (item.indexdef.indexOf(' pkey ') > -1) {
                  extra[name].primary = true;
                }
                var index = item.indexdef.indexOf(' UNIQUE ') > -1 ? 'unique' : 'index';
                extra[name][index] = others.length ? others : true;
              });

              return _context.abrupt('return', think.extend(schema, extra));

            case 15:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    }));

    function getSchema(_x2) {
      return _ref.apply(this, arguments);
    }

    return getSchema;
  }();
  /**
   * start transaction
   * @return {Promise} []
   */


  _class.prototype.startTrans = function startTrans() {
    if (this.transTimes === 0) {
      this.transTimes++;
      return this.execute('BEGIN');
    }
    this.transTimes++;
    return _promise2.default.resolve();
  };
  /**
  * parse where item
  * @param  {String} key []
  * @param  {Mixed} val []
  * @return {String}     []
  */


  _class.prototype.parseWhereItem = function parseWhereItem(key, val) {
    var _this2 = this;

    // {id: null}
    if (val === null) {
      return key + ' IS NULL';
    }
    // {id: {'<': 10, '>': 1}}
    else if (think.isObject(val)) {
        var logic = this.getLogic(val);
        var result = [];
        for (var opr in val) {
          var nop = opr.toUpperCase();
          nop = this.comparison[nop] || nop;
          var parsedValue = this.parseValue(val[opr]);
          //{id: {IN: [1, 2, 3]}}
          if (think.isArray(parsedValue)) {
            result.push(key + ' ' + nop + ' (' + parsedValue.join(', ') + ')');
          } else if (parsedValue === 'null') {
            result.push(key + ' ' + (nop === '!=' ? 'IS NOT NULL' : 'IS NULL'));
          } else {
            result.push(key + ' ' + nop + ' ' + parsedValue);
          }
        }
        return result.join(' ' + logic + ' ');
      }
      // where({id: [1, 2, 3]})
      else if (think.isArray(val)) {
          var flag = think.isNumber(val[0]) || think.isNumberString(val[0]);
          if (flag) {
            flag = val.every(function (item) {
              return think.isNumber(item) || think.isNumberString(item);
            });
            if (flag) {
              return key + ' IN ( ' + val.join(', ') + ' )';
            }
          }
        } else {
          return key + ' = ' + this.parseValue(val);
        }

    var whereStr = '';
    var data = void 0;
    if (think.isString(val[0])) {
      (function () {
        var val0 = val[0].toUpperCase();
        val0 = _this2.comparison[val0] || val0;
        // compare
        if (/^(=|!=|>|>=|<|<=)$/.test(val0)) {
          if (val[1] === null) {
            whereStr += key + ' ' + (val[0] === '!=' ? 'IS NOT NULL' : 'IS NULL');
          } else {
            whereStr += key + ' ' + val0 + ' ' + _this2.parseValue(val[1]);
          }
        }
        // like or not like
        else if (/^(NOT\s+LIKE|LIKE|NOT\s+ILIKE|ILIKE)$/.test(val0)) {
            if (think.isArray(val[1])) {
              //get like logic, default is OR
              var likeLogic = _this2.getLogic(val[2], 'OR');
              var like = val[1].map(function (item) {
                return key + ' ' + val0 + ' ' + _this2.parseValue(item);
              }).join(' ' + likeLogic + ' ');
              whereStr += '(' + like + ')';
            } else {
              whereStr += key + ' ' + val0 + ' ' + _this2.parseValue(val[1]);
            }
          }
          // exp
          else if (val0 === 'EXP') {
              whereStr += '(' + key + ' ' + val[1] + ')';
            }
            // in or not in
            else if (val0 === 'IN' || val0 === 'NOT IN') {
                if (val[2] === 'exp') {
                  whereStr += key + ' ' + val0 + ' ' + val[1];
                } else {
                  if (think.isString(val[1])) {
                    val[1] = val[1].split(',');
                  }
                  if (!think.isArray(val[1])) {
                    val[1] = [val[1]];
                  }
                  val[1] = _this2.parseValue(val[1]);
                  if (val[1].length === 1) {
                    whereStr += key + (val0 === 'IN' ? ' = ' : ' != ') + val[1];
                  } else {
                    whereStr += key + ' ' + val0 + ' (' + val[1].join(',') + ')';
                  }
                }
              }
              //between
              else if (val0 === 'BETWEEN' || val0 === 'NOT BETWEEN') {
                  data = think.isString(val[1]) ? val[1].split(',') : val[1];
                  if (!think.isArray(data)) {
                    data = [val[1], val[2]];
                  }
                  whereStr += ' (' + key + ' ' + val0 + ' ' + _this2.parseValue(data[0]);
                  whereStr += ' AND ' + _this2.parseValue(data[1]) + ')';
                } else {
                  throw new Error(think.locale('WHERE_CONDITION_INVALID', key, (0, _stringify2.default)(val)));
                }
      })();
    } else {

      var length = val.length;
      var _logic = this.getLogic(val[length - 1], '');
      if (_logic) {
        length--;
      } else {
        _logic = 'AND';
      }
      var _result = [];
      for (var i = 0; i < length; i++) {
        var isArr = think.isArray(val[i]);
        data = isArr ? val[i][1] : val[i];
        var exp = ((isArr ? val[i][0] : '') + '').toUpperCase();
        if (exp === 'EXP') {
          _result.push('(' + key + ' ' + data + ')');
        } else {
          var op = isArr ? this.comparison[val[i][0].toUpperCase()] || val[i][0] : '=';
          _result.push('(' + key + ' ' + op + ' ' + this.parseValue(data) + ')');
        }
      }
      whereStr = _result.join(' ' + _logic + ' ');
    }
    return whereStr;
  };

  _class.prototype.quoteKey = function quoteKey(key) {
    if (key === undefined || think.isEmpty(key)) {
      return '';
    }
    if (think.isNumber(key) || think.isNumberString(key)) {
      return key;
    }
    if (/.*\(.*\)/.test(key)) {
      return key;
    }
    if (/(.*[a-z0-9]+)(\")([a-z0-9]+.*)/i.test(key)) {
      return key.replace(/(.*[a-z0-9]+)(\")([a-z0-9]+.*)/i, '\"$1\"\"$3\"');
    } else {
      return '"' + key + '"';
    }
  };
  /**
   * parse key
   * @param  {String} key []
   * @return {String}     []
   */


  _class.prototype.parseKey = function parseKey(key) {
    var _this3 = this;

    if (key === undefined) {
      return '';
    }
    if (think.isNumber(key) || think.isNumberString(key)) {
      return key;
    }
    key = key.trim();
    if (think.isEmpty(key)) {
      return '';
    }
    // EXAMPLE: 'user_age(birthday)' or 'user_age(birthday) AS age' 
    if (/.*\(.*\)/.test(key)) {
      return key;
    }
    var isDistinct = false;
    if (/DISTINCT (.*)/i.test(key)) {
      isDistinct = true;
      key = key.replace(/DISTINCT (.*)/i, '$1');
    }
    if (/.*\..*/.test(key)) {
      var k = key.split('.'),
          j = [];
      k.forEach(function (i) {
        var tmp = _this3.quoteKey(i.replace(/^[\"]+|[\"]+$/g, ''));
        j.push('' + tmp);
      });
      key = j.join('.');
    } else {
      key = this.quoteKey(key.replace(/^[\"]+|[\"]+$/g, ''));
    }
    return '' + (isDistinct ? 'DISTINCT ' : '') + key;
  };
  /**
   * parse group
   * @param  {String} group []
   * @return {String}       []
   */


  _class.prototype.parseGroup = function parseGroup(group) {
    if (think.isEmpty(group)) {
      return '';
    }
    if (think.isString(group)) {
      //group may be `date_format(create_time,'%Y-%m-%d')`
      if (group.indexOf('(') !== -1) {
        return ' GROUP BY ' + group;
      }
      group = group.split(/\s*,\s*/);
    }
    var result;

    if (think.isArray(group)) {
      result = group.map(function (item) {
        item = item.replace(/[\"]/g, '');
        var type = '',
            regexp = /(.*) (ASC|DESC)/i,
            matches = item.match(regexp);

        if (matches !== null) {
          type = ' ' + matches[2];
          item = item.replace(regexp, '$1');
        }

        if (item.indexOf('.') === -1) {
          return '"' + item + '"' + type;
        } else {
          item = item.split('.');
          return '"' + item[0] + '"."' + item[1] + '"' + type;
        }
      });

      return ' GROUP BY ' + result.join(', ');
      /**
       * Example: { 'name': 'DESC' } || { 'name': -1 }
       */
    } else if (think.isObject(group)) {
      result = [];

      for (var key in group) {
        var type = group[key],
            matches = void 0;

        key = key.replace(/[\"]/g, '');

        if (think.isString(type)) {
          matches = type.match(/.*(ASC|DESC)/i);
        }

        if (matches) {
          type = ' ' + matches[1];
        } else if (think.isNumber(type) || think.isNumberString(type)) {
          type = parseInt(type) === -1 ? ' DESC' : ' ASC';
        }

        if (key.indexOf('.') === -1) {
          result.push('"' + key + '"' + type);
        } else {
          key = key.split('.');

          result.push('"' + key[0] + '"."' + key[1] + '"' + type);
        }
      }

      return ' GROUP BY ' + result.join(', ');
    } else {
      /** Unknown format: */
    }
  };
  /**
   * parse limit
   * @param  {String} limit []
   * @return {String}       []
   */


  _class.prototype.parseLimit = function parseLimit(limit) {
    if (think.isEmpty(limit)) {
      return '';
    }
    if (think.isNumber(limit)) {
      return ' LIMIT ' + limit;
    }
    if (think.isString(limit)) {
      limit = limit.split(/\s*,\s*/);
    }
    if (limit[1]) {
      return ' LIMIT ' + (limit[1] | 0) + ' OFFSET ' + (limit[0] | 0);
    }
    return ' LIMIT ' + (limit[0] | 0);
  };
  /**
   * parse value
   * @param  {Mixed} value []
   * @return {Mixed}       []
   */


  _class.prototype.parseValue = function parseValue(value) {
    var _this4 = this;

    if (think.isString(value)) {
      value = 'E\'' + this.escapeString(value) + '\'';
    } else if (think.isArray(value)) {
      if (/^exp$/.test(value[0])) {
        value = value[1];
      } else {
        value = value.map(function (item) {
          return _this4.parseValue(item);
        });
      }
    } else if (think.isBoolean(value)) {
      value = value ? 'true' : 'false';
    } else if (value === null) {
      value = 'null';
    }
    return value;
  };
  /**
   * query string
   * @param  string str
   * @return promise
   */


  _class.prototype.query = function query(sql) {
    var _this5 = this;

    this.sql = sql;
    return think.await(sql, function () {
      return _this5.socket(sql).query(sql).then(function (data) {
        return _this5.bufferToString(data.rows);
      });
    });
  };
  /**
   * execute sql
   * @param  {String} sql []
   * @return {}     []
   */


  _class.prototype.execute = function execute(sql) {
    var _this6 = this;

    this.sql = sql;
    var insertInto = 'insert into ';
    var prefix = sql.slice(0, insertInto.length).toLowerCase();
    var isInsert = false;
    if (prefix === insertInto) {
      sql += ' RETURNING id';
      isInsert = true;
    }
    return this.socket(sql).execute(sql).then(function (data) {
      if (isInsert) {
        _this6.lastInsertId = data.rows[0].id;
      }
      return data.rowCount || 0;
    });
  };

  return _class;
}(_base2.default);

exports.default = _class;
