// thinkjs/lib/util/watch_compile.js
'use strict';

exports.__esModule = true;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * watch compile
 */
var _class = function () {
  /**
   * constructor
   * @param  {Array} args []
   * @return {}         []
   */

  /**
   * compiled error files
   * @type {Array}
   */
  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    this.compiledMtime = {};
    this.compiledErrorFiles = [];
    this.allowFileExt = ['.js', '.ts'];

    this.init.apply(this, arguments);
  }
  /**
   * init
   * @param  {String} srcPath []
   * @param  {String} outPath []
   * @param  {Boolean} log     []
   * @return {}         []
   */

  /**
   * allow file ext in src path
   * @type {Array}
   */

  /**
   * store compiled files last mtime
   * @type {Object}
   */


  _class.prototype.init = function init(srcPath, outPath) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    var callback = arguments[3];

    this.srcPath = _path2.default.normalize(srcPath);
    this.outPath = _path2.default.normalize(outPath);
    this.options = options;
    this.callback = callback;
  };
  /**
   * compile single file
   * @param  {String} file     []
   * @param  {Boolean} onlyCopy []
   * @return {}          []
   */


  _class.prototype.compileFile = function compileFile(file, onlyCopy) {
    var filePath = '' + this.srcPath + think.sep + file;
    var content = _fs2.default.readFileSync(filePath, 'utf8');

    //when get file content empty, maybe file is locked
    if (!content) {
      return;
    }
    // only copy file content
    if (onlyCopy) {
      var saveFilepath = '' + this.outPath + think.sep + file;
      think.mkdir(_path2.default.dirname(saveFilepath));
      _fs2.default.writeFileSync(saveFilepath, content);
      return;
    }

    try {
      if (this.options.type === 'ts') {
        this.compileByTypeScript(content, file);
      } else {
        this.compileByBabel(content, file);
      }
      return true;
    } catch (e) {

      think.log(function (colors) {
        return colors.red('compile file ' + file + ' error');
      }, 'COMPILE');
      think.log(e);

      e.message = 'Compile Error: ' + e.message;
      think.compileError = e;
    }
    return false;
  };
  /**
   * get relative path
   * @param  {String} file []
   * @return {String}      []
   */


  _class.prototype.getRelationPath = function getRelationPath(file) {
    //use dirname to resolve file path in source-map-support
    //so must use dirname in here
    var pPath = _path2.default.dirname(this.outPath + think.sep + file);
    return _path2.default.relative(pPath, this.srcPath + think.sep + file);
  };
  /**
   * typescript compile
   * @return {} []
   */


  _class.prototype.compileByTypeScript = function compileByTypeScript(content, file) {
    var ts = require('typescript');
    var startTime = Date.now();
    var diagnostics = [];
    var output = ts.transpileModule(content, {
      compilerOptions: {
        module: ts.ModuleKind.CommonJS,
        target: ts.ScriptTarget.ES6,
        experimentalDecorators: true,
        emitDecoratorMetadata: true,
        allowSyntheticDefaultImports: true,
        sourceMap: true
      },
      fileName: file,
      reportDiagnostics: !!diagnostics
    });
    ts.addRange(diagnostics, output.diagnostics);

    //has error
    if (diagnostics.length) {
      var firstDiagnostics = diagnostics[0];

      var _firstDiagnostics$fil = firstDiagnostics.file.getLineAndCharacterOfPosition(firstDiagnostics.start),
          line = _firstDiagnostics$fil.line,
          character = _firstDiagnostics$fil.character;

      var message = ts.flattenDiagnosticMessageText(firstDiagnostics.messageText, '\n');
      throw new Error(message + ' on Line ' + (line + 1) + ', Character ' + character);
    }
    if (this.options.log) {
      think.log('Compile file ' + file, 'TypeScript', startTime);
    }

    file = this.replaceExtName(file, '.js');
    var sourceMap = JSON.parse(output.sourceMapText);
    sourceMap.sources[0] = this.getRelationPath(file);
    sourceMap.sourcesContent = [content];
    //file value must be equal sources values
    sourceMap.file = sourceMap.sources[0];
    delete sourceMap.sourceRoot;
    this.compileByBabel(output.outputText, file, true, sourceMap);
  };
  /**
   * babel compile
   * @return {} []
   */


  _class.prototype.compileByBabel = function compileByBabel(content, file, logged, orginSourceMap) {
    var startTime = Date.now();
    var relativePath = this.getRelationPath(file);
    //babel not export default property
    //so can not use `import babel from 'babel-core'`
    var babel = require('babel-core');
    var data = babel.transform(content, {
      filename: file,
      presets: [].concat(this.options.presets || ['es2015-loose', 'stage-1']),
      plugins: [].concat(this.options.plugins || ['transform-runtime']),
      sourceMaps: true,
      sourceFileName: relativePath
    });
    if (!logged && this.options.log) {
      think.log('Compile file ' + file, 'Babel', startTime);
    }
    think.mkdir(_path2.default.dirname('' + this.outPath + think.sep + file));
    var basename = _path2.default.basename(file);
    var prefix = '//# sourceMappingURL=';
    if (data.code.indexOf(prefix) === -1) {
      data.code = data.code + '\n' + prefix + basename + '.map';
    }
    _fs2.default.writeFileSync('' + this.outPath + think.sep + file, data.code);
    var sourceMap = data.map;
    //file value must be equal sources values
    sourceMap.file = sourceMap.sources[0];
    if (orginSourceMap) {
      sourceMap = this.mergeSourceMap(orginSourceMap, sourceMap);
    }
    _fs2.default.writeFileSync('' + this.outPath + think.sep + file + '.map', (0, _stringify2.default)(sourceMap, undefined, 4));
  };
  /**
   * merge source map
   * @param  {String} content        []
   * @param  {Object} orginSourceMap []
   * @param  {Object} sourceMap      []
   * @return {}                []
   */


  _class.prototype.mergeSourceMap = function mergeSourceMap(orginSourceMap, sourceMap) {
    var _require = require('source-map'),
        SourceMapGenerator = _require.SourceMapGenerator,
        SourceMapConsumer = _require.SourceMapConsumer;

    sourceMap.file = sourceMap.file.replace(/\\/g, '/');
    sourceMap.sources = sourceMap.sources.map(function (filePath) {
      return filePath.replace(/\\/g, '/');
    });
    var generator = SourceMapGenerator.fromSourceMap(new SourceMapConsumer(sourceMap));
    generator.applySourceMap(new SourceMapConsumer(orginSourceMap));
    sourceMap = JSON.parse(generator.toString());

    return sourceMap;
  };
  /**
   * src file is deleted, but app file also exist
   * then delete app file
   * @return {} []
   */


  _class.prototype.getSrcDeletedFiles = function getSrcDeletedFiles(srcFiles, appFiles) {
    var _this = this;

    var srcFilesWithoutExt = srcFiles.map(function (item) {
      return _this.replaceExtName(item);
    });
    return appFiles.filter(function (file) {
      var extname = _path2.default.extname(file);
      if (_this.allowFileExt.indexOf(extname) === -1) {
        return;
      }
      var fileWithoutExt = _this.replaceExtName(file);
      //src file not exist
      if (srcFilesWithoutExt.indexOf(fileWithoutExt) === -1) {
        var filepath = _this.outPath + think.sep + file;
        if (think.isFile(filepath)) {
          _fs2.default.unlinkSync(filepath);
        }
        return true;
      }
    }).map(function (file) {
      return _this.outPath + think.sep + file;
    });
  };
  /**
   * replace filepath extname
   * @param  {String} filepath []
   * @param  {String} extname  []
   * @return {String}          []
   */


  _class.prototype.replaceExtName = function replaceExtName(filepath) {
    var extname = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

    return filepath.replace(/\.\w+$/, extname);
  };
  /**
   * compile
   * @return {} []
   */


  _class.prototype.compile = function compile(once) {
    var _this2 = this;

    var files = think.getFiles(this.srcPath, true);
    var appFiles = think.getFiles(this.outPath, true);
    var changedFiles = this.getSrcDeletedFiles(files, appFiles);

    if (think.compileError && !this.compiledErrorFiles.length) {
      think.compileError = null;
    }

    files.forEach(function (file) {
      var extname = _path2.default.extname(file);
      //if is not js file, only copy
      if (_this2.allowFileExt.indexOf(extname) === -1) {
        _this2.compileFile(file, true);
        return;
      }
      var mTime = _fs2.default.statSync('' + _this2.srcPath + think.sep + file).mtime.getTime();
      var outFile = '' + _this2.outPath + think.sep + file;

      //change extname to .js.
      //in typescript, file extname is .ts
      outFile = _this2.replaceExtName(outFile, '.js');

      if (think.isFile(outFile)) {
        var outmTime = _fs2.default.statSync(outFile).mtime.getTime();
        //if compiled file mtime is after than source file, return
        if (outmTime >= mTime) {
          return;
        }
      }
      if (!_this2.compiledMtime[file] || mTime > _this2.compiledMtime[file]) {
        var ret = _this2.compileFile(file);
        if (ret) {
          changedFiles.push(outFile);
        }

        _this2.compiledMtime[file] = mTime;

        var index = _this2.compiledErrorFiles.indexOf(file);
        if (ret) {
          if (index > -1) {
            _this2.compiledErrorFiles.splice(index, 1);
          }
        } else if (ret === false) {
          if (index === -1) {
            _this2.compiledErrorFiles.push(file);
          }
        }
      }
    });
    //notify auto reload service to clear file cache
    if (changedFiles.length && this.callback) {
      this.callback(changedFiles);
    }
    if (!once) {
      setTimeout(this.compile.bind(this), 100);
    }
  };
  /**
   * run
   * @return {} []
   */


  _class.prototype.run = function run() {
    this.compile();
  };
  /**
   * compile
   * @return {} []
   */


  _class.compile = function compile(srcPath, outPath) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

    var instance = new this(srcPath, outPath, options);
    instance.compile(true);
  };

  return _class;
}();

exports.default = _class;
