// thinkjs/lib/adapter/socket/memcache.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _base = require('./base.js');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * memcache socket
 */
var _class = function (_Base) {
  (0, _inherits3.default)(_class, _Base);

  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    return (0, _possibleConstructorReturn3.default)(this, _Base.apply(this, arguments));
  }

  /**
   * inti
   * @param  {Object} config []
   * @return {}        []
   */
  _class.prototype.init = function init() {
    var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _Base.prototype.init.call(this, config);

    this.config = think.extend({
      host: '127.0.0.1',
      port: 11211,
      username: '',
      password: ''
    }, config);
  };
  /**
   * get socket connect str
   * @return {String} []
   */


  _class.prototype.getSocketStr = function getSocketStr(protocal) {
    var str = '';
    var config = this.config;
    if (config.username) {
      str += config.username;
    }
    if (config.password) {
      str += ':' + config.password;
    }
    if (str) {
      str += '@';
    }
    str += config.host + ':' + config.port;
    if (protocal) {
      return 'memcache://' + str;
    }
    return str;
  };
  /**
   * get connection
   * @return {Promise} []
   */


  _class.prototype.getConnection = function () {
    var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
      var memjs;
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!this.connection) {
                _context.next = 2;
                break;
              }

              return _context.abrupt('return', this.connection);

            case 2:
              _context.next = 4;
              return think.npm('memjs');

            case 4:
              memjs = _context.sent;

              this.connection = memjs.Client.create(this.getSocketStr(), {
                logger: {
                  log: function log() {}
                }
              });

              this.logConnect(this.getSocketStr(true), 'memcache');

              return _context.abrupt('return', this.connection);

            case 8:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    }));

    function getConnection() {
      return _ref.apply(this, arguments);
    }

    return getConnection;
  }();
  /**
   * wrap method
   * @param  {String}    name []
   * @param  {} args []
   * @return {Promise}         []
   */


  _class.prototype.wrap = function () {
    var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(name) {
      for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      var connection, deferred, callback, err;
      return _regenerator2.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return this.getConnection();

            case 2:
              connection = _context2.sent;
              deferred = think.defer();

              callback = function callback(err, data) {
                return err ? deferred.reject(err) : deferred.resolve(data && data.toString());
              };

              if (args.length === 1) {
                args.push(callback);
              } else {
                args = [args[0], args[1], callback, args[2]];
              }
              connection[name].apply(connection, args);
              err = new Error(this.getSocketStr(true));
              return _context2.abrupt('return', think.error(deferred.promise, err));

            case 9:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, this);
    }));

    function wrap(_x2, _x3) {
      return _ref2.apply(this, arguments);
    }

    return wrap;
  }();
  /**
   * get data
   * @param  {String} key []
   * @return {Promise}     []
   */


  _class.prototype.get = function get(key) {
    return this.wrap('get', key);
  };
  /**
   * set data 
   * @param {String} key     []
   * @param {String} value   []
   * @param {Number} timeout []
   */


  _class.prototype.set = function set(key, value) {
    var timeout = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.config.timeout;

    return this.wrap('set', key, value, timeout);
  };
  /**
   * delete data
   * @param  {String} key []
   * @return {Promise}     []
   */


  _class.prototype.delete = function _delete(key) {
    return this.wrap('delete', key);
  };
  /**
   * increment
   * @param  {String} key     []
   * @param  {Number} amount  []
   * @param  {Number} timeout []
   * @return {Promise}         []
   */


  _class.prototype.increment = function increment(key, amount) {
    var timeout = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.config.timeout;

    return this.wrap('increment', key, amount, timeout);
  };
  /**
   * decrement
   * @param  {String} key     []
   * @param  {Number} amount  []
   * @param  {Number} timeout []
   * @return {Promise}         []
   */


  _class.prototype.decrement = function decrement(key, amount) {
    var timeout = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.config.timeout;

    return this.wrap('decrement', key, amount, timeout);
  };

  return _class;
}(_base2.default);

exports.default = _class;
