// thinkjs/lib/adapter/socket/redis.js
'use strict';

exports.__esModule = true;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _base = require('./base.js');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Redis socket class
 */
var _class = function (_Base) {
  (0, _inherits3.default)(_class, _Base);

  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    return (0, _possibleConstructorReturn3.default)(this, _Base.apply(this, arguments));
  }

  /**
   * init
   * @param  {Object} config []
   * @return {}        []
   */
  _class.prototype.init = function init() {
    var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _Base.prototype.init.call(this, config);

    this.config = think.extend({
      port: 6379,
      host: '127.0.0.1',
      password: ''
    }, config);
  };
  /**
   * connect redis
   * @return {Promise} []
   */


  _class.prototype.getConnection = function () {
    var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
      var _this2 = this;

      var redis, config, str;
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!this.connection) {
                _context.next = 2;
                break;
              }

              return _context.abrupt('return', this.connection);

            case 2:
              _context.next = 4;
              return think.npm('redis');

            case 4:
              redis = _context.sent;
              config = this.config;
              str = 'redis://' + config.host + ':' + config.port;
              return _context.abrupt('return', think.await(str, function () {

                var deferred = think.defer();
                var connection = redis.createClient(config.port, config.host, config);
                if (config.password) {
                  connection.auth(config.password, function () {});
                }
                connection.on('connect', function () {
                  _this2.connection = connection;
                  _this2.logConnect(str, 'redis');
                  deferred.resolve(connection);
                });
                connection.on('error', function (err) {
                  _this2.close();
                  _this2.logConnect(str, 'redis');
                  deferred.reject(err);
                });
                var err = new Error(str);
                return think.error(deferred.promise, err);
              }));

            case 8:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    }));

    function getConnection() {
      return _ref.apply(this, arguments);
    }

    return getConnection;
  }();
  /**
   * add event
   * @param  {String}   event    []
   * @param  {Function} callback []
   * @return {}            []
   */


  _class.prototype.on = function on(event, callback) {
    return this.getConnection().then(function (connection) {
      connection.on(event, callback);
    });
  };
  /**
   * wrap
   * @param  {String}    name []
   * @param  {Array} data []
   * @return {Promise}         []
   */


  _class.prototype.wrap = function () {
    var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(name) {
      for (var _len = arguments.length, data = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        data[_key - 1] = arguments[_key];
      }

      var fn;
      return _regenerator2.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return this.getConnection();

            case 2:
              fn = think.promisify(this.connection[name], this.connection);
              return _context2.abrupt('return', think.error(fn.apply(undefined, data)));

            case 4:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, this);
    }));

    function wrap(_x2, _x3) {
      return _ref2.apply(this, arguments);
    }

    return wrap;
  }();
  /**
   * get data
   * @param  {String} name []
   * @return {Promise}      []
   */


  _class.prototype.get = function get(name) {
    return this.wrap('get', name);
  };
  /**
   * set data
   * @param {String} name    []
   * @param {String} value   []
   * @param {Number} timeout []
   */


  _class.prototype.set = function set(name, value) {
    var timeout = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.config.timeout;

    var setP = [this.wrap('set', name, value)];
    if (timeout) {
      setP.push(this.expire(name, timeout));
    }
    return _promise2.default.all(setP);
  };
  /**
   * set data expire
   * @param  {String} name    []
   * @param  {Number} timeout []
   * @return {Promise}         []
   */


  _class.prototype.expire = function expire(name, timeout) {
    return this.wrap('expire', name, timeout);
  };
  /**
   * delete data
   * @param  {String} name []
   * @return {Promise}      []
   */


  _class.prototype.delete = function _delete(name) {
    return this.wrap('del', name);
  };
  /**
   * close socket connection
   * @return {} []
   */


  _class.prototype.close = function close() {
    if (this.connection) {
      this.connection.end();
      this.connection = null;
    }
  };

  return _class;
}(_base2.default);

exports.default = _class;
