// thinkjs/lib/model/_base.js
'use strict';

exports.__esModule = true;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _maxSafeInteger = require('babel-runtime/core-js/number/max-safe-integer');

var _maxSafeInteger2 = _interopRequireDefault(_maxSafeInteger);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _think_validate = require('../core/think_validate.js');

var _think_validate2 = _interopRequireDefault(_think_validate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var forceNewNum = 1;

/**
 * base model class
 */

var _class = function (_think$base) {
  (0, _inherits3.default)(_class, _think$base);

  function _class() {
    (0, _classCallCheck3.default)(this, _class);
    return (0, _possibleConstructorReturn3.default)(this, _think$base.apply(this, arguments));
  }

  /**
   * init
   * @param  {} name   []
   * @param  {} config []
   * @return {}        []
   */
  _class.prototype.init = function init() {
    var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};


    var options = {
      pk: 'id', //primary key
      name: '', //model name
      tablePrefix: undefined, //table prefix
      tableName: '', //table name, without prefix
      /**
       * schema
       * {
       *   name: {
       *     type: 'string',
       *     required: true,
       *     default: '',
       *     unique: true,
       *     primary: true,
       *     auto_increment: true
       *   }
       * }
       */
      schema: {}, //table schema
      /**
       * table indexes
       * {
       *   name: {name: 1, $unique: true}, //single indexes
       *   test1: {name1: 1, title1: 1, $unique: 1}, //compound indexes
       * }
       * @type {Object}
       */
      indexes: {}
      //readonlyFields: []// readonly fields
    };
    //if is set in subclass, can't be override
    for (var key in options) {
      if (this[key] === undefined) {
        this[key] = options[key];
      }
    }

    if (think.isObject(name)) {
      config = name;
      name = '';
    }

    config = think.parseConfig(true, config);

    //change property name `name`
    if (config.name && !config.database) {
      config.database = config.name;
      delete config.name;
      think.log('db.name is deprecated, use db.database instead', 'WARNING');
    }

    if (config.pwd && !config.password) {
      config.password = config.pwd;
      delete config.pwd;
      think.log('db.pwd is deprecated, use db.password instead', 'WARNING');
    }

    //check property name `fields`
    if (!think.isEmpty(this.fields)) {
      this.schema = this.fields;
      delete this.fields;
      think.log('fields property is deprecated, use schema instead', 'WARNING');
    }

    this.config = config;
    this._db = null;
    this._data = {};
    this._options = {};

    //model name
    if (name) {
      this.name = name;
    }
    // get table prefix from config
    if (this.config.prefix && this.tablePrefix === undefined) {
      this.tablePrefix = this.config.prefix;
    }
  };
  /**
   * get model instance
   * @param  {String} name    [model name]
   * @param  {Object} options [options]
   * @return {Object}         []
   */


  _class.prototype.model = function model(name, options, module) {
    if (think.isString(options) && think.module.indexOf(options) > -1) {
      module = options;
      options = {};
    }
    module = module || this.parseModuleFromPath();
    if (think.isString(options)) {
      options = { type: options };
    }
    options = think.extend({}, this.config, options);
    return think.model(name, options, module);
  };
  /**
   * get table prefix
   * @return {String} []
   */


  _class.prototype.getTablePrefix = function getTablePrefix() {
    return this.tablePrefix || '';
  };
  /**
   * get db instance
   * @return {Object} []
   */


  _class.prototype.db = function db() {
    var forceNew = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

    // set db
    if (think.isObject(forceNew)) {
      this._db = forceNew;
      return this;
    }
    if (this._db && !forceNew && !this.config.parser) {
      return this._db;
    }
    var DB = think.adapter('db', this.config.type || 'mysql');
    var config = this.config;
    if (forceNewNum > _maxSafeInteger2.default - 1) {
      forceNewNum = 1;
    }
    if (forceNew) {
      config = think.extend({}, config, { forceNewNum: forceNewNum++ });
    }
    this._db = new DB(config);
    return this._db;
  };
  /**
   * get model name
   * @return {String} []
   */


  _class.prototype.getModelName = function getModelName() {
    if (this.name) {
      return this.name;
    }
    var filename = this.__filename || __filename;
    var last = filename.lastIndexOf(think.sep);
    this.name = filename.substr(last + 1, filename.length - last - 4);
    return this.name;
  };
  /**
   * get table name
   * @return {String} []
   */


  _class.prototype.getTableName = function getTableName() {
    if (!this.tableName) {
      this.tableName = this.getModelName();
    }
    return this.getTablePrefix() + this.tableName;
  };
  /**
   * set cache options
   * @param  {String} key     []
   * @param  {Number} timeout []
   * @return {}         []
   */


  _class.prototype.cache = function cache(key) {
    var timeout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.config.cache.timeout;

    if (key === undefined) {
      return this;
    }
    var options = void 0;
    if (!think.isObject(key)) {
      if (think.isNumber(key)) {
        timeout = key;
        key = '';
      }
      options = think.extend({}, this.config.cache, { key: key, timeout: timeout });
    } else {
      options = key;
    }
    this._options.cache = options;
    return this;
  };
  /**
   * set limit options
   * @param  {Number} offset []
   * @param  {Number} length []
   * @return {}        []
   */


  _class.prototype.limit = function limit(offset, length) {
    if (offset === undefined) {
      return this;
    }
    if (think.isArray(offset)) {
      length = offset[1] || length;
      offset = offset[0];
    }
    offset = Math.max(parseInt(offset) || 0, 0);
    if (length) {
      length = Math.max(parseInt(length) || 0, 0);
    }
    this._options.limit = [offset, length];
    return this;
  };
  /**
   * set page options
   * @param  {Number} page     []
   * @param  {} listRows []
   * @return {}          []
   */


  _class.prototype.page = function page(_page) {
    var listRows = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.config.nums_per_page;

    if (_page === undefined) {
      return this;
    }
    if (think.isArray(_page)) {
      listRows = _page[1] || listRows;
      _page = _page[0];
    }
    _page = Math.max(parseInt(_page) || 1, 1);
    listRows = Math.max(parseInt(listRows) || 10, 1);
    this._options.limit = [listRows * (_page - 1), listRows];
    return this;
  };
  /**
   * set where options
   * @return {} []
   */


  _class.prototype.where = function where(_where) {
    if (!_where) {
      return this;
    }
    if (think.isString(_where)) {
      _where = { _string: _where };
    }
    var options = this._options;
    if (options.where && think.isString(options.where)) {
      options.where = { _string: options.where };
    }
    options.where = think.extend({}, options.where, _where);
    return this;
  };
  /**
   * set field options
   * @param  {String} field   []
   * @param  {Boolean} reverse []
   * @return {}         []
   */


  _class.prototype.field = function field(_field) {
    var reverse = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

    if (!_field) {
      return this;
    }
    if (think.isString(_field)) {
      if (_field.indexOf(')') === -1) {
        _field = _field.split(/\s*,\s*/);
      }
    }
    this._options.field = _field;
    this._options.fieldReverse = reverse;
    return this;
  };
  /**
   * set field reverse
   * @param  {String} field [field list]
   * @return {Object}       []
   */


  _class.prototype.fieldReverse = function fieldReverse(field) {
    return this.field(field, true);
  };
  /**
   * set table name
   * @param  {String} table []
   * @return {}       []
   */


  _class.prototype.table = function table(_table, hasPrefix) {
    if (!_table) {
      return this;
    }
    _table = _table.trim();
    //table is sql, `SELECT * FROM`
    if (_table.indexOf(' ') > -1) {
      hasPrefix = true;
    }
    this._options.table = hasPrefix ? _table : this.getTablePrefix() + _table;
    return this;
  };
  /**
   * union options
   * @param  {} union []
   * @param  {} all   []
   * @return {}       []
   */


  _class.prototype.union = function union(_union) {
    var all = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

    if (!_union) {
      return this;
    }
    if (!this._options.union) {
      this._options.union = [];
    }
    this._options.union.push({
      union: _union,
      all: all
    });
    return this;
  };
  /**
   * .join({
   *   'xxx': {
   *     join: 'left',
   *     as: 'c',
   *     on: ['id', 'cid']
   *   }
   * })
   * @param  {[type]} join [description]
   * @return {[type]}      [description]
   */


  _class.prototype.join = function join(_join) {
    if (!_join) {
      return this;
    }
    if (!this._options.join) {
      this._options.join = [];
    }
    if (think.isArray(_join)) {
      this._options.join = this._options.join.concat(_join);
    } else {
      this._options.join.push(_join);
    }
    return this;
  };
  /**
   * set order options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.order = function order(value) {
    this._options.order = value;
    return this;
  };
  /**
   * set table alias
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.alias = function alias(value) {
    this._options.alias = value;
    return this;
  };
  /**
   * set having options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.having = function having(value) {
    this._options.having = value;
    return this;
  };
  /**
   * set group options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.group = function group(value) {
    this._options.group = value;
    return this;
  };
  /**
   * set lock options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.lock = function lock(value) {
    this._options.lock = value;
    return this;
  };
  /**
   * set auto options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.auto = function auto(value) {
    this._options.auto = value;
    return this;
  };
  /**
   * set filter options
   * @param  {String} value []
   * @return {}       []
   */


  _class.prototype.filter = function filter(value) {
    this._options.filter = value;
    return this;
  };
  /**
   * set distinct options
   * @param  {String} data []
   * @return {}      []
   */


  _class.prototype.distinct = function distinct(data) {
    this._options.distinct = data;
    if (think.isString(data)) {
      this._options.field = data;
    }
    return this;
  };
  /**
   * set explain
   * @param  {Boolean} explain []
   * @return {}         []
   */


  _class.prototype.explain = function explain(_explain) {
    this._options.explain = _explain;
    return this;
  };
  /**
   * options filter
   * @param  {Object} options []
   * @return {}         []
   */


  _class.prototype.optionsFilter = function optionsFilter(options) {
    return options;
  };
  /**
   * data filter
   * @param  {Object} data []
   * @return {}      []
   */


  _class.prototype.dataFilter = function dataFilter(data) {
    return data;
  };
  /**
   * before add
   * @param  {Object} data []
   * @return {}      []
   */


  _class.prototype.beforeAdd = function beforeAdd(data, options, schema) {
    var _this2 = this;

    //for addMany invoked
    if (think.isArray(data)) {
      return data.map(function (item) {
        return _this2.beforeAdd(item, options);
      });
    }

    var ret = {};
    var extRet = {};
    schema = schema || this.schema;
    //fields in schema
    for (var field in schema) {
      var fieldSchema = schema[field];
      var _default = fieldSchema.default;
      //default value is setted
      if (!think.isTrueEmpty(_default)) {
        ret[field] = {
          value: data[field],
          default: _default
        };
      } else {
        if (this._isSubSchema(fieldSchema)) {
          extRet[field] = this.beforeAdd(data[field] || {}, options, fieldSchema);
        }
      }
    }
    for (var _field2 in data) {
      if (!ret[_field2] && !extRet[_field2]) {
        ret[_field2] = {
          value: data[_field2]
        };
      }
    }
    ret = _think_validate2.default.values(ret);
    if (!think.isEmpty(extRet)) {
      ret = think.extend(ret, extRet);
    }
    return ret;
  };
  /**
   * check is sub schema
    // meta: {
    //   createAt: {
    //     default: ()=>new Date()
    //   },
    //   updateAt: {
    //     default: ()=>new Date()
    //   }
    // }
   * @param  {Mixed}  schema []
   * @return {Boolean}        []
   */


  _class.prototype._isSubSchema = function _isSubSchema(schema) {
    if (!schema || !think.isObject(schema)) {
      return false;
    }
    var keys = (0, _keys2.default)(schema);
    return keys.length && keys.every(function (key) {
      return think.isObject(schema[key]);
    });
  };
  /**
   * after add
   * @param  {} data []
   * @return {}      []
   */


  _class.prototype.afterAdd = function afterAdd(data) {
    return data;
  };
  /**
   * before delete
   */


  _class.prototype.beforeDelete = function beforeDelete(options) {
    return options;
  };
  /**
   * after delete
   * @param  {Mixed} data []
   * @return {}      []
   */


  _class.prototype.afterDelete = function afterDelete(data) {
    return data;
  };
  /**
   * before update
   * @param  {Mixed} data []
   * @return {}      []
   */


  _class.prototype.beforeUpdate = function beforeUpdate(data, options, schema) {
    var _this3 = this;

    //check property readonlyFields
    if (!think.isEmpty(this.readonlyFields)) {
      (function () {
        var ret = {};
        _this3.readonlyFields.forEach(function (item) {
          ret[item] = { readonly: true };
        });
        delete _this3.readonlyFields;
        _this3.schema = think.extend(ret, _this3.schema);
        think.log('readonlyFields property is deprecated, use schema[field].readonly instead', 'WARNING');
      })();
    }

    var ret = {};
    var extRet = {};
    schema = schema || this.schema;

    for (var field in data) {
      var fieldSchema = schema[field];
      if (!fieldSchema) {
        ret[field] = { value: data[field] };
      } else {
        if (this._isSubSchema(fieldSchema)) {
          var result = this.beforeUpdate(data[field] || {}, options, fieldSchema);
          if (!think.isEmpty(result)) {
            extRet[field] = result;
          }
        } else if (!fieldSchema.readonly) {
          ret[field] = { value: data[field] };
        }
      }
    }

    for (var _field3 in schema) {
      var _fieldSchema = schema[_field3];
      var _default = _fieldSchema.default;
      if (!think.isTrueEmpty(_default) && !_fieldSchema.readonly && _fieldSchema.update) {
        ret[_field3] = {
          value: data[_field3],
          default: _default
        };
      } else if (this._isSubSchema(_fieldSchema)) {
        var _result = this.beforeUpdate(data[_field3] || {}, options, _fieldSchema);
        if (!think.isEmpty(_result)) {
          extRet[_field3] = _result;
        }
      }
    }
    ret = _think_validate2.default.values(ret);
    if (!think.isEmpty(extRet)) {
      ret = think.extend(ret, extRet);
    }
    return ret;
  };
  /**
   * after update
   * @param  {} data    []
   * @param  {} options []
   * @return {}         []
   */


  _class.prototype.afterUpdate = function afterUpdate(data) {
    return data;
  };
  /**
   * before find
   */


  _class.prototype.beforeFind = function beforeFind(options) {
    return options;
  };
  /**
   * after find
   * @return {} []
   */


  _class.prototype.afterFind = function afterFind(data) {
    return data;
  };
  /**
   * before select
   */


  _class.prototype.beforeSelect = function beforeSelect(options) {
    return options;
  };
  /**
   * after select
   * @param  {Mixed} result []
   * @return {}        []
   */


  _class.prototype.afterSelect = function afterSelect(data) {
    return data;
  };
  /**
   * set data
   * @param  {Mixed} data []
   * @return {}      []
   */


  _class.prototype.data = function data(_data) {
    if (_data === true) {
      return this._data;
    }
    this._data = _data;
    return this;
  };
  /**
   * set options
   * @param  {Mixed} options []
   * @return {}         []
   */


  _class.prototype.options = function options(_options) {
    if (!_options) {
      return this._options;
    }
    this._options = _options;
    //page to limit
    if (_options.page) {
      this.page(_options.page);
    }
    return this;
  };
  /**
   * close db socket
   * @return {} []
   */


  _class.prototype.close = function close() {
    if (this._db) {
      this._db.close();
      this._db = null;
    }
  };

  return _class;
}(think.base);

exports.default = _class;
